Test/tutorial/ProvenTerminatingFixpoint.dfy

```diff
@@ -0,0 +1,259 @@
+// RUN: %dafny_0 /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+/// # Fixpoint combinators with provable termination guarantees
+///
+/// *Mikael Mayer*, 2022/09/22
+///
+/// Fixpoints are functions combinators such that, if they exist,
+/// then we don't need to define recursion or mutual recursion at all,
+/// only use fixpoints when we need recursion.
+/// Let's start by illustrating an example
+
+module {:options "-functionSyntax:4"} Fixpoints {
+
+/// We first define the usual Fibonacci function
+/// To make Dafny happy, we need to provide a (normally implicit) decreases clause,
+/// and Dafny will verify that 0 <= n-1 < n and 0 <= n-2 < n when n > 2
+
+  function Fib1(n: nat): nat
+    decreases n
+  {
+    if n <= 1 then n else Fib1(n-1) + Fib1(n-2)
+  }
+
+/// This is nice, but mathematically speaking, we are still defining a function in terms of itself,
+/// which is not obvious.
+/// Besides a while loop and a stack to implement recursive functions,
+/// there is also another way , using fixpoint combinators.
+///
+/// A fixpoint combinator typically takes a 'template function' that, if provided with the final function and an input,
+/// execute the body of the function we want where they replace recursive calls with the provided final functions.
+/// We can write a fibonacci function using a combinator like this:
+
+  function Fib2(n: nat): nat {
+    var comb :=
+      (f: nat -> nat, n': nat) => if n' <= 1 then n' as nat else f(n'-1) + f(n'-2);
+    fixpoint2(comb, n)
+  }
+
+/// The signature of this fixpoint combinator could typically be:
+
+  function fixpoint2(comb: (nat -> nat, nat) -> nat, u: nat): nat
+
+/// However, any attempt at writing the body of fixpoint will fail, because Dafny will always check for
+/// termination.
+  {
+    var callback := (u': nat) => fixpoint2(comb, u');
+    comb(callback, u)
+  }
+
+/// So let's fix that. Fix, we need to realize that, in order for fixpoint to finish, we need to
+/// that u' is less than u. But doing so changes the type of the first argument of comb from nat -> nat
+/// to nat --> nat (see [arrow subset types](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-arrow-subset-types))
+/// Let's do the change
+
+  function fixpoint3(comb: (nat --> nat, nat) -> nat, u: nat): nat
+    decreases u
+  {
+    var callback := (u': nat) requires u' < u => fixpoint3(comb, u');
+    comb(callback, u)
```

Comment:
I'm a bit confused about the names `comb` and `callback`.  Isn't `fixpoint` itself the combinator?